junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ForkJoinPoolHierarchicalTestExecutorService.java

```diff
@@ -155,29 +157,34 @@ public void invokeAll(List<? extends TestTask> tasks) {
 			new ExclusiveTask(tasks.get(0)).execSync();
 			return;
 		}
-		Deque<ExclusiveTask> nonConcurrentTasks = new LinkedList<>();
+		Deque<ExclusiveTask> isolatedTasks = new LinkedList<>();
+		Deque<ExclusiveTask> sameThreadTasks = new LinkedList<>();
 		Deque<ExclusiveTask> concurrentTasksInReverseOrder = new LinkedList<>();
-		forkConcurrentTasks(tasks, nonConcurrentTasks, concurrentTasksInReverseOrder);
-		executeNonConcurrentTasks(nonConcurrentTasks);
+		forkConcurrentTasks(tasks, isolatedTasks, sameThreadTasks, concurrentTasksInReverseOrder);
+		executeSync(sameThreadTasks);
 		joinConcurrentTasksInReverseOrderToEnableWorkStealing(concurrentTasksInReverseOrder);
+		executeSync(isolatedTasks);
 	}
 
-	private void forkConcurrentTasks(List<? extends TestTask> tasks, Deque<ExclusiveTask> nonConcurrentTasks,
-			Deque<ExclusiveTask> concurrentTasksInReverseOrder) {
+	private void forkConcurrentTasks(List<? extends TestTask> tasks, Deque<ExclusiveTask> isolatedTasks,
+			Deque<ExclusiveTask> sameThreadTasks, Deque<ExclusiveTask> concurrentTasksInReverseOrder) {
 		for (TestTask testTask : tasks) {
 			ExclusiveTask exclusiveTask = new ExclusiveTask(testTask);
-			if (testTask.getExecutionMode() == CONCURRENT) {
-				exclusiveTask.fork();
-				concurrentTasksInReverseOrder.addFirst(exclusiveTask);
+			if (testTask.getResourceLock() instanceof GlobalReadWriteLock) {
+				isolatedTasks.add(exclusiveTask);
+			}
+			else if (testTask.getExecutionMode() == SAME_THREAD) {
+				sameThreadTasks.add(exclusiveTask);
 			}
 			else {
-				nonConcurrentTasks.add(exclusiveTask);
+				exclusiveTask.fork();
+				concurrentTasksInReverseOrder.addFirst(exclusiveTask);
 			}
 		}
 	}
 
-	private void executeNonConcurrentTasks(Deque<ExclusiveTask> nonConcurrentTasks) {
-		for (ExclusiveTask task : nonConcurrentTasks) {
+	private void executeSync(Deque<ExclusiveTask> tasks) {
```

Comment:
I guess we are using `Deque` for consistency because here, we treat it more as a list instead of consuming the elements.

Reply:
Yes, solely for consistency, any `Iterable` would do